### PR TITLE
Create SpecialRemovePII (T7286: phase two and phase three)

### PR DESCRIPTION
**Pull request by Universal-Omega**

### CODE REVIEW

**Review by dmehus — COMMENTED** [review]
:+1: LGTM so far. It's weird to see no comma in the "services" array because it only has one item, and the last item has no comma after it. :P

File: includes/RemovePIIJob.php
Comment (Southparkfan) [inline includes/RemovePIIJob.php:70]: Shouldn't this be configured in LocalSettings.php and/or through extension hooks? Not only to support third-party usage, but to avoid making lots of commits to this extension as well.

File: includes/RemovePIIJob.php
Comment (Southparkfan) [inline includes/RemovePIIJob.php:436]: Can we switch this to a $dbw->delete() action?

File: includes/RemovePIIJob.php
Comment (Southparkfan) [inline includes/RemovePIIJob.php:444]: Why are we suppressing warnings? That is an indication of a fault in the deletion logic...

If you need to suppress warnings, please use the best practices from https://github.com/wikimedia/at-ease.

File: includes/SpecialRemovePII.php
Comment (Southparkfan) [inline includes/SpecialRemovePII.php:122]: Duplicating the global rename logic into this extension may be useful to reduce the need for a second manual action (renaming users), but is prone to errors. Do you think Miraheze can keep up with maintaining this piece of code?

File: .github/workflows/continuousIntegration.yml
Comment (Southparkfan) [inline .github/workflows/continuousIntegration.yml:13]: Please include PHP 7.4 as well, I hope to migrate to PHP 7.4 later this year (PHP 8 support is not present in MediaWiki yet).

File: includes/RemovePIIJob.php
Comment (Universal-Omega) [reply includes/RemovePIIJob.php:436]: I know it's probably possible, but I'm not exactly sure how to do it with the Join sub query.

File: includes/RemovePIIJob.php
Comment (Universal-Omega) [reply includes/RemovePIIJob.php:70]: Yes I had planned to add hook support at least and make this a configuration, but was planning to split that into another PR. But can do that now if you think I should.

File: includes/RemovePIIJob.php
Comment (Universal-Omega) [reply includes/RemovePIIJob.php:444]: Because when the revision is deleted before the page it gives undefined index 'content' warning. And I can't do revision deletion after because then the page no longer exists in page table and the join doesn't work. But I can switch to atEase.

File: includes/SpecialRemovePII.php
Comment (Universal-Omega) [reply includes/SpecialRemovePII.php:122]: I think it can be maintained. It's only duplicates a small part of it and CentralAuth would only change something like that on 1.35->1.36 upgrades for example (major upgrades) so checking that before upgrading would make it easy to maintain. I did this because eventually I want to combine it into one step (the RemovePII  and Renameuser at the exact same time to prevent every logging the renames)

File: .github/workflows/continuousIntegration.yml
Comment (Universal-Omega) [reply .github/workflows/continuousIntegration.yml:13]: Technically this file just says edited because it was added in a merge of the master branch to this one, so I'll do that after merging this so I don't merge conflict things further.

**Comment by paladox** [comment]
test....

File: includes/RemovePIIJob.php
Comment (Universal-Omega) [reply includes/RemovePIIJob.php]: Changed to atEase

File: includes/SpecialRemovePII.php
Comment (Universal-Omega) [inline includes/SpecialRemovePII.php:141]: Just adding a couple personal notes. Can we create a new class in this extension, which extends this `GlobalRenameUserLogger` class and override the `log()` function to `return false;` to never log renames done here?

File: includes/RemovePIIJob.php
Comment (Universal-Omega) [inline includes/RemovePIIJob.php:488]: If we can make the rename never log then we can likely combine this all into one step. Adapt the above code to be the first thing to run, once all that's done automatically start the rename job. This would definitely simplify the whole thing and make the old username never shown in any logs.

File: includes/SpecialRemovePII.php
Comment (Universal-Omega) [reply includes/SpecialRemovePII.php:122]: Also just to add, with #15, this has an additional usage to never log renames done with removePII to never reveal the old name elsewhere.

File: includes/RemovePIIJob.php
Comment (supertassu) [inline includes/RemovePIIJob.php:67]: `DB_PRIMARY`. Jobs don't support DI yet, but when they do this should be replaced with the DBLoadBalancerFactory service

File: includes/RemovePIIJob.php
Comment (supertassu) [inline includes/RemovePIIJob.php:48]: consider using UserIdentities if possible, and UserFactory if not

File: includes/RemovePIIJob.php
Comment (supertassu) [inline includes/RemovePIIJob.php:349]: use TitleFactory instead

File: includes/RemovePIIJob.php
Comment (supertassu) [inline includes/RemovePIIJob.php:445]: these will likely be completely refactored in the next few releases, I'm not aware of what's replacing them but if they can be used please do, otherwise use the factories

File: includes/RemovePIIJob.php
Comment (supertassu) [inline includes/RemovePIIJob.php:457]: this is the last statement in the foreach body, did you mean `break;`?

File: includes/RemovePIIJob.php
Comment (supertassu) [inline includes/RemovePIIJob.php:438]: please add a comment explaining what this is supposed to do, not immediately clear to me when looking

**Review by supertassu — CHANGES_REQUESTED** [review]
First pass, see inline comments. In addition, the special page needs checks to make sure that users without relevant permissions (`centralauth-oversight`/`hideuser`/whatever) can't see suppressed (global) users.

File: includes/SpecialRemovePII.php
Comment (Universal-Omega) [reply includes/SpecialRemovePII.php:107]: Reason why I didn't is the whole extension doesn't require those extensions.

File: includes/SpecialRemovePII.php
Comment (Universal-Omega) [reply includes/SpecialRemovePII.php]: This has been done.

**Comment by Universal-Omega** [comment]
> First pass, see inline comments. In addition, the special page needs checks to make sure that users without relevant permissions (`centralauth-oversight`/`hideuser`/whatever) can't see suppressed (global) users.

Done with f3cf2eb

File: includes/SpecialRemovePII.php
Comment (supertassu) [inline includes/SpecialRemovePII.php:66]: Are there other possible actions?

File: phplint.sh
Comment (supertassu) [inline phplint.sh:4]: consider using standard mediawiki tools (codesniffer, phan) instead

**Review by supertassu — COMMENTED** [review]
haven't tested, but looks good aside from the few inline comments

File: phplint.sh
Comment (Universal-Omega) [reply phplint.sh:4]: This would likely be done as an entirely different PR, as it should be done for all our extensions as well probably.

File: includes/SpecialRemovePII.php
Comment (Universal-Omega) [reply includes/SpecialRemovePII.php:66]: There is the RenameUser action.

File: includes/RemovePIIJob.php
Comment (supertassu) [inline includes/RemovePIIJob.php:349]: this (and all other database calls) should include `__METHOD__` for easier debugging. depending on the table sizes also consider chunking with `waitForReplicas()` calls in between

Comment (Universal-Omega) [reply includes/RemovePIIJob.php:349]: Oh right thanks!

Comment (Universal-Omega) [reply includes/RemovePIIJob.php:349]: Also do you mean `MediaWikiServices::getInstance()->getDBLoadBalancerFactory()->waitForReplication()`? Or is there actually a `waitForReplicas()` (if so I've just not heard of it but I might be wrong)

**Comment by Universal-Omega** [comment]
Testing on test seems like this all works correctly as it should. It can probably be deployed now if wanted.

File: includes/RemovePIIJob.php
Comment (supertassu) [reply includes/RemovePIIJob.php]: Yeah, oops, it's `waitForReplication()`, but if you operate on some other database than the specific wikis main one you need to pass the database name to the function. I'm used to working with CA which aliases it as [waitForReplicas()](https://github.com/wikimedia/mediawiki-extensions-CentralAuth/blob/28e1ec49febb43d5ea79e9266bcd79f0ee150af1/includes/CentralAuthUtilityService.php#L92), sorry for the wrong name.

**Comment by Universal-Omega** [comment]
Will be testing and hopefully deploy this PR soon. This comment is just for my personal use so I don't forget to take another look at some things.

Things to note for a potential follow-up PR, not this one as this one currently maintains the same behavior as the script but can still possibly be improved:

* If the account is not created locally, can we automatically do `CentralAuthServices::getForcedLocalCreationService()->attemptAutoCreateLocalUserFromName( $formData['oldname'] );` in the `onSubmit()` function?
* Can we combine all steps into one simple process? Currently the only thing that is blocking this is the creation of a local rename log, if we can suppress that log everything else can become simplified into one step.
* Can we convert `$tableDeletions` and `$tableUpdates` to a config to be set elsewhere? Maybe a way to simplify that so it's not so huge also.

Something to consider **before** merging this PR:
* Should we use `OutputPage::disallowUserJs()` to disallow any user javascript on the special page, exactly like on Special:Preferences?

**Comment by Universal-Omega** [comment]
This now seems to work 100% as it should.

**Pull request merged**
